docs: design document for multi-cluster service support (#11318)
rook/rook: design/ceph/submariner-globalnet.md

- Create ServiceExport resource for each mon and OSD service.
```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1
kind: ServiceExport
  name: <name>
  namespace: rook-ceph
```
- Here, the ServiceExport resource name should be the name of the service that should be exported.

## Implementation Details

- For each mon and OSDs service:
    - Create a corresponding `ServiceExport` resource.
    - Verify the status of the `ServiceExport` from the `ServiceExport` resource status. Sample Status:
    ```yaml
    Status:
    Conditions:
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Awaiting sync of the ServiceImport to the broker
        Reason:                AwaitingSync
        Status:                False
        Type:                  Valid
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Service was successfully synced to the broker
        Reason:
        Status:                True
        Type:                  Valid
    ```
    - Obtain the global IP for the exported service by issuing a DNS query to `<service>.<ns>.svc.clusterset.local`.

>     - Use this IP in the `--public-addr` flag when creating the mon or OSD deployment.

Reviewer: How about adding this bullet as well?
- The OSD will bind to the pod IP with the new OSD flag `--public-bind-addr`.

  name: <name>
  namespace: rook-ceph
```
- Here, the ServiceExport resource name should be the name of the service that should be exported.

## Implementation Details

- For each mon and OSDs service:
    - Create a corresponding `ServiceExport` resource.
    - Verify that status of the `ServiceExport` should be `true`. Sample Status:
    ```yaml
    Status:
    Conditions:
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Awaiting sync of the ServiceImport to the broker
        Reason:                AwaitingSync
        Status:                False
        Type:                  Valid
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Service was successfully synced to the broker
        Reason:
        Status:                True
        Type:                  Valid
    ```
    - Obtain the global IP for the exported service by issuing a DNS query to `<service>.<ns>.svc.clusterset.local`.
    - Use this IP in the `--public-addr` flag when creating the mon or OSD deployment.
    - The OSD will bind to the POD IP with the flag `--public-bind-addr`
    - Ensure that mon endpoints configMap has the global IPs.

> ## Upgrade

Reviewer 2: ```suggestion
## Upgrade
```
question: 
why do we have this section?

Author: Mostly to cover the case when user has a cluster running and then decides to enable multiClusterService.  Maybe I need to think of a better name.

> # Submariner Globalnet support in Rook

Reviewer 3: ```suggestion
# Multi-cluster Service support in Rook
```

## Use case
As a rook-ceph user, I should be able to setup mirroing across multiple clusters with overlapping networks in order to safegurd my data in case of any disaster.

## Background

> [Submariner](https://submariner.io/getting-started/) provides cross-cluster connectivity between two kubernetes clusters when they have different CIDRs. Many a times customers don't have control over the network configuration of the clusters, resulting in clusters with overlapping CIDRs (ServiceCIDR and ClusterCIDR). For such scenarios, Submariner provides another tool called [Globalnet](https://submariner.io/getting-started/architecture/globalnet/). It provides cross-cluster connectivity from pods to remote services using their global IPs.

Reviewer 3: ```suggestion
Sometimes users wish to connect two Kubernetes clusters into a single logical cluster. Often, both clusters may have a standardized install with overlapping CIDRs (Service CIDR and/or Pod network CIDR). The Kubernetes "sig-multicluster" community SIG defines a Multi-Cluster Services (MCS) API for providing cross-cluster connectivity from pods to remote services using global IPs.

> In order to support kubernetes clusters connected using Submariner-Globalnet, Rook needs to use the global IPs provided by Globalnet instead of the service cluster IP.

Reviewer 3: ```suggestion
In order to support Kubernetes clusters connected by an MCS API-compatible application, Rook needs to use "clusterset" IPs instead of the Services' cluster IPs.
```

> 

Reviewer 3: Add background here that addresses that we don't want to hard-code support for only Submariner. 

Submariner uses the upstream Multi-Cluster Services (MCS) API which is being defined by the K8s SIG “mcs-api”. This is also associated with the K8s SIG community “sig-multicluster”.
MCS API SIG GitHub - https://github.com/kubernetes-sigs/mcs-api

Reviewer: While we don't want the implementation specific to Submariner, how about adding details about what we have observed about Submariner behavior that is important to consider for any implementation? For example:
- The traffic in-cluster to mons and OSDs is observed to be routed directly to the mons or OSDs. This is very important so traffic is not routed through a single bottleneck.
- The traffic across cluster is all routed through the submariner gateway. This known to be a bottleneck, but is necessary for the tunnel.

## Prerequisite

> - The peer clusters should be connected using Submariner.
> - Globalnet controller should be running on both the clusters.
> (Refer Submariner [documentation](https://submariner.io/getting-started/quickstart/) for installation steps on different environments)

Reviewer 3: ```suggestion
- peer clusters should be connected using an MCS API-compatible application.
```

> ## Service of OSD pods

Reviewer: In addition to the OSD services, mention that the OSD daemon will be aware of the IP with the flags:
```
--public-addr=<externalIP>
--public-bind-addr=<podIP>
```
The latter flag still needs to be implemented by ceph.

Author: I've added this point in the `Implementation Details` section below. Let me know that would suffice.

> - Implement k8s service for each osd pod:

Reviewer 3: ```suggestion
- Implement k8s service for each OSD pod:
```

> ## Service of OSD pods
> - Implement k8s service for each osd pod:

Reviewer 3: Explaining why this is needed will help any upstream users who aren't aware of what we have previously researched.

```suggestion
## Service of OSD pods
For scenarios like RBD mirroring, peers need direct access to Ceph OSDs. Each OSD will have to have a standard ClusterIP Service created for it to allow this. The OSD Service will be created only when multi-cluster networking support is enabled.

The reference implementation used for development of this feature is (Submariner)[link]. In the reference implementation, it is important for Services to be of `type` ClusterIP. Headless Services don't have internal routing between OSDs local to a cluster -- or any other Ceph daemons local to the cluster.
```

@travisn should we always enable OSD services or only for MCS?

```yaml
apiVersion: v1
  kind: Service
  metadata:
    labels:
      app: rook-ceph-osd
      app.kubernetes.io/component: cephclusters.ceph.rook.io
      app.kubernetes.io/created-by: rook-ceph-operator
      app.kubernetes.io/instance: <osd-id>
      app.kubernetes.io/managed-by: rook-ceph-operator
      app.kubernetes.io/name: ceph-osd
      app.kubernetes.io/part-of: rook-ceph
      ceph_daemon_id: <osd-id>
      ceph_daemon_type: osd
      osd: <osd-id>
      osd_cluster: rook-ceph
      rook.io/operator-namespace: rook-ceph
      rook_cluster: rook-ceph
    name: rook-ceph-mon-<id>
    namespace: rook-ceph
  spec:
    selector:
      app: rook-ceph-osd
      ceph_daemon_id: <osd-id>
      osd: <osd-id>
      osd_cluster: rook-ceph
      rook_cluster: rook-ceph
    type: ClusterIP

> ```

Reviewer 3: I think defining something as simple as the Service here is probably a bit unnecessary. I would probably omit this.

> ## API Changes

Reviewer 3: Nit: want space between these?

> - Add a new `provider` type of `submariner-globalnet` under `spec.network`

Reviewer 2: ```suggestion
- Add a new `provider` type of `submariner-globalnet` under `spec.network`
```
@sp98 is this setting for ceph cluster cr? if yes can we mention `under `spec.network in cephCluster cr`.

``` yaml
spec:
    network:
        ### submariner-globalnet should be used when kubernetes clusters have overlapping CIDRs
        provider: submariner-globalnet

> ```

Reviewer 3: I've mentioned several times that we have had bad experiences with Rook in the past from hard-coding providers. I think it's okay to mention in the design doc that we are developing the feature with submariner, we should be using the Multi-Cluster Services (MCS) API which is being defined by the K8s SIG “mcs-api”.

We should also consider the possibility that users wish to use multus and MCS at the same time. Instead of using the existing `network.provider` here. I think we should make a `network.multiCluster.enabled` field. Making the `enabled` nested will allow us to add other multi-cluster-related fields to the spec in the future if needed.

https://github.com/kubernetes-sigs/mcs-api

Author: @BlaineEXE  Thanks, I'll update the spec to use mcs.

## Service Export
- `ServiceExport` CR is used to specify which services should be exposed across all the clusters in the cluster set.

> - The exported service becomes accessible as `<service>.<ns>.svc.clusterset.local`

Reviewer 3: ```suggestion
- The exported service becomes accessible as `<service>.<ns>.svc.clusterset.local`.
```

> - Create ServiceExport resource for each mon and osd service.

Reviewer 3: ```suggestion
- Create ServiceExport resource for each mon and OSD service.
```

```yaml

> apiVersion: multicluster.x-k8s.io/v1alpha1

Author: I'll check and get back to you on this.

> ```yaml
> apiVersion: multicluster.x-k8s.io/v1alpha1
>   kind: ServiceExport
>     name: <name>
>     namespace: rook-ceph
> ```

Reviewer 3: Spacing off

```suggestion
```yaml
apiVersion: multicluster.x-k8s.io/v1alpha1
kind: ServiceExport
  name: <name>
  namespace: rook-ceph
```
```

Author: fixed.

- Here, the ServiceExport resource name should be the name of the service that should be exported.

> 

Reviewer 3: Nit: Why an inconsistent number of spaces between sections? There are 3 here but 2 between others and one between others. Generally, 1 space between sections is what Rook uses.

Author: fixed.

## Implementation Details
- Select `submariner-globalnet` as `spec.network.provider`
- For each mon and osd service:
    - Create a corresponding `ServiceExport` resource.
    - Verify the status of the `ServiceExport` from the `ServiceExport` resource status. Sample Status:
    ```yaml
    Status:
    Conditions:
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Awaiting sync of the ServiceImport to the broker
        Reason:                AwaitingSync
        Status:                False
        Type:                  Valid
        Last Transition Time:  2020-12-01T12:35:32Z
        Message:               Service was successfully synced to the broker

>         Reason:

Reviewer 3: Why is `Reason` empty here? Isn't this a required field that the export should have?

Author: Most probably because its in a `success` state. I'll check again and get back to you on this.

        Status:                True
        Type:                  Valid
    ```

>     - Obtain the external IP/ global IP from the exported service.

Reviewer 3: ```suggestion
    - Obtain the global IP for the exported service by issuing a DNS query to `<service>.<ns>.svc.clusterset.local`.
```

Reviewer: I believe we can just get this from the `externalIP` field on the service, no need for a DNS query?

Reviewer 3: The `ExternalIP` resource is a submariner thing, not an MCS API thing, so we can't rely on that. The MCS way of doing it is the DNS query.

    - Use this IP in the `--public-addr` flag when creating the mon or osd deployment

>     - Ensure that mon endpoints configMap have the global IPs

Reviewer 3: ```suggestion
    - Ensure that mon endpoints configMap has the global IPs
```
